include/exadg/solvers_and_preconditioners/multigrid/multigrid_preconditioner_base.cpp

```diff
@@ -309,6 +312,54 @@ MultigridPreconditionerBase<dim, Number>::check_levels(std::vector<MGLevelInfo>
   }
 }
 
+#if DEAL_II_VERSION_GTE(10, 0, 0)
+
+/**
+ * A class to use for the deal.II coarsening functionality, where we try to
+ * balance the mesh coarsening with a minimum granularity and the number of
+ * partitions on coarser levels.
+ */
+template<int dim, int spacedim = dim>
+class BalancedGranularityPartitionPolicy : public RepartitioningPolicyTools::Base<dim, spacedim>
+{
+public:
+  BalancedGranularityPartitionPolicy(types::global_cell_index const n_cells_fine,
+                                     unsigned int const             n_mpi_processes)
+    : n_cells_per_level{n_cells_fine}, n_mpi_processes_per_level{n_mpi_processes}
+  {
+  }
+
+  virtual LinearAlgebra::distributed::Vector<double>
+  partition(Triangulation<dim, spacedim> const & tria_coarse_in) const override
+  {
+    types::global_cell_index const n_cells = tria_coarse_in.n_global_active_cells();
+
+    // TODO: We hard-code a grain-size limit of 200 cells per processor
+    // (assuming linear finite elements and typical behavior of
+    // supercomputers), but we also do not want to coarsen by more than a
+    // factor of 8 to not create too many MPI messages.
+    unsigned int const grain_size_limit =
+      std::min(200U, 8 * n_cells / n_mpi_processes_per_level.back() + 1);
+
+    RepartitioningPolicyTools::MinimalGranularityPolicy<dim, spacedim> partitioning_policy(
+      grain_size_limit);
+    LinearAlgebra::distributed::Vector<double> const partitions =
+      partitioning_policy.partition(tria_coarse_in);
+
+    // The vector 'partitions' contains the partition numbers. To get the
+    // number of partitions, we take the infinity norm.
+    n_mpi_processes_per_level.push_back(static_cast<unsigned int>(partitions.linfty_norm()) + 1);
+    n_cells_per_level.push_back(n_cells);
+    return partitions;
+  }
+
+private:
+  mutable std::vector<types::global_cell_index> n_cells_per_level;
```

Comment:
Why do you need this variable? It is private, but only filled in the function `partition()` without being used by this class or another one if I see it correctly.

Reply:
It is not used in the final version; however, it might be that we use more clever strategies in the future, so I left it in when I derived the algorithm in the present form. If you want, we can remove it now and bring it back in case we do a more refined strategy.

Reply:
It does not help right now in my opinion.